src/server/search_family.cc

```diff
@@ -0,0 +1,244 @@
+#include "base/logging.h"
+#include "core/search/search.h"
+#include "facade/reply_builder.h"
+#include "server/command_registry.h"
+#include "server/conn_context.h"
+#include "server/container_utils.h"
+#include "server/engine_shard_set.h"
+#include "server/transaction.h"
+
+extern "C" {
+#include "redis/listpack.h"
+#include "redis/object.h"
+};
+
+namespace dfly {
+
+using namespace std;
+using namespace facade;
+
+namespace {
+
+string_view SdsToSafeSv(sds str) {
+  return str != nullptr ? string_view{str, sdslen(str)} : ""sv;
+}
+
+using DocumentData = absl::flat_hash_map<std::string, std::string>;
+using SerializedDocument = pair<std::string /*key*/, DocumentData>;
+using Query = search::AstExpr;
+
+struct BaseAccessor : public search::HSetAccessor {
+  using FieldConsumer = search::HSetAccessor::FieldConsumer;
+
+  virtual DocumentData Serialize() const = 0;
+};
+
+struct ListPackAccessor : public BaseAccessor {
+  using LpPtr = uint8_t*;
+
+  ListPackAccessor(LpPtr ptr) : lp_{ptr} {
+  }
+
+  bool Check(FieldConsumer f, string_view active_field) const override {
+    uint8_t intbuf[LP_INTBUF_SIZE];
+
+    if (!active_field.empty()) {
+      return f(container_utils::LpFind(lp_, active_field, intbuf).value_or(""));
+    }
+
+    uint8_t* fptr = lpFirst(lp_);
+    DCHECK_NE(fptr, nullptr);
+
+    while (fptr) {
+      fptr = lpNext(lp_, fptr);  // skip key
+      string_view v = container_utils::LpGetView(fptr, intbuf);
+      fptr = lpNext(lp_, fptr);
+
+      if (f(v))
+        return true;
+    }
+
+    return false;
+  }
+
+  DocumentData Serialize() const override {
+    uint8_t intbuf[2][LP_INTBUF_SIZE];
+    DocumentData out{};
+
+    uint8_t* fptr = lpFirst(lp_);
+    DCHECK_NE(fptr, nullptr);
+
+    while (fptr) {
+      string_view k = container_utils::LpGetView(fptr, intbuf[0]);
+      fptr = lpNext(lp_, fptr);  // skip key
+      string_view v = container_utils::LpGetView(fptr, intbuf[1]);
+      fptr = lpNext(lp_, fptr);
+
+      out[k] = v;
+    }
+
+    return out;
+  }
+
+ private:
+  LpPtr lp_;
+};
+
+struct StringMapAccessor : public BaseAccessor {
+  StringMapAccessor(StringMap* hset) : hset_{hset} {
+  }
+
+  bool Check(FieldConsumer f, string_view active_field) const override {
+    if (!active_field.empty()) {
+      return f(SdsToSafeSv(hset_->Find(active_field)));
+    }
+
+    for (const auto& [k, v] : *hset_) {
+      if (f(SdsToSafeSv(v)))
+        return true;
+    }
+
+    return false;
+  }
+
+  DocumentData Serialize() const override {
+    DocumentData out{};
+    for (const auto& [k, v] : *hset_)
+      out[SdsToSafeSv(k)] = SdsToSafeSv(v);
+    return out;
+  }
+
+ private:
+  StringMap* hset_;
+};
+
+unique_ptr<BaseAccessor> GetAccessor(const OpArgs& op_args, const PrimeValue& pv) {
+  if (pv.Encoding() == kEncodingListPack) {
+    auto ptr = reinterpret_cast<ListPackAccessor::LpPtr>(pv.RObjPtr());
+    return make_unique<ListPackAccessor>(ptr);
+  } else {
+    DCHECK_EQ(pv.Encoding(), kEncodingStrMap2);
```

Comment:
I guess now this `DCHECK` is not needed, as it's done inside?
(But feel free to keep if you think it's still helpful, even if only slightly)